File: sdks/rust/tests/graph_tests.rs
//! Tests for graph operations in the Rust SDK

use vectorizer_sdk::*;

#[tokio::test]
async fn test_list_graph_nodes() {
    let client = VectorizerClient::new_default().unwrap();
    
    // This test requires a collection with graph enabled
    // For now, we just test that the method exists and can be called
    let result = client.list_graph_nodes("test_collection").await;
    
    // If collection doesn't exist, we expect an error
    // If it exists, we should get a valid response
    match result {
        Ok(response) => {
            assert!(response.count >= 0);
            assert_eq!(response.nodes.len(), response.count);
        }
        Err(_) => {
            // Collection doesn't exist or graph not enabled - this is expected in test environment
        }
    }
}

#[tokio::test]
async fn test_get_graph_neighbors() {
    let client = VectorizerClient::new_default().unwrap();
    
    let result = client.get_graph_neighbors("test_collection", "test_node").await;
    
    match result {
        Ok(response) => {
            assert!(!response.neighbors.is_empty() || response.neighbors.is_empty());
        }
        Err(_) => {
            // Expected if collection/node doesn't exist
        }
    }
}

#[tokio::test]
async fn test_find_related_nodes() {
    let client = VectorizerClient::new_default().unwrap();
    
    let request = FindRelatedRequest {
        max_hops: Some(2),
        relationship_type: Some("SIMILAR_TO".to_string()),
    };
    
    let result = client.find_related_nodes("test_collection", "test_node", request).await;
    
    match result {
        Ok(response) => {
            assert!(response.related.len() >= 0);
        }
        Err(_) => {
            // Expected if collection/node doesn't exist
        }
    }
}

#[tokio::test]
async fn test_find_graph_path() {
    let client = VectorizerClient::new_default().unwrap();
    
    let request = FindPathRequest {
        collection: "test_collection".to_string(),
        source: "node1".to_string(),
        target: "node2".to_string(),
    };
    
    let result = client.find_graph_path(request).await;
    
    match result {
        Ok(response) => {
            assert!(response.found || !response.found);
            if response.found {
                assert!(!response.path.is_empty());
            }
        }
        Err(_) => {
            // Expected if collection/nodes don't exist
        }
    }
}

#[tokio::test]
async fn test_create_graph_edge() {
    let client = VectorizerClient::new_default().unwrap();
    
    let request = CreateEdgeRequest {
        collection: "test_collection".to_string(),
        source: "node1".to_string(),
        target: "node2".to_string(),
        relationship_type: "SIMILAR_TO".to_string(),
        weight: Some(0.85),
    };
    
    let result = client.create_graph_edge(request).await;
    
    match result {
        Ok(response) => {
            assert!(response.success);
            assert!(!response.edge_id.is_empty());
        }
        Err(_) => {
            // Expected if collection/nodes don't exist
        }
    }
}

#[tokio::test]
async fn test_list_graph_edges() {
    let client = VectorizerClient::new_default().unwrap();
    
    let result = client.list_graph_edges("test_collection").await;
    
    match result {
        Ok(response) => {
            assert!(response.count >= 0);
            assert_eq!(response.edges.len(), response.count);
        }
        Err(_) => {
            // Expected if collection doesn't exist
        }
    }
}

#[tokio::test]
async fn test_discover_graph_edges() {
    let client = VectorizerClient::new_default().unwrap();
    
    let request = DiscoverEdgesRequest {
        similarity_threshold: Some(0.7),
        max_per_node: Some(10),
    };
    
    let result = client.discover_graph_edges("test_collection", request).await;
    
    match result {
        Ok(response) => {
            assert!(response.success);
            assert!(response.edges_created >= 0);
        }
        Err(_) => {
            // Expected if collection doesn't exist
        }
    }
}

#[tokio::test]
async fn test_get_graph_discovery_status() {
    let client = VectorizerClient::new_default().unwrap();
    
    let result = client.get_graph_discovery_status("test_collection").await;
    
    match result {
        Ok(response) => {
            assert!(response.total_nodes >= 0);
            assert!(response.nodes_with_edges >= 0);
            assert!(response.total_edges >= 0);
            assert!(response.progress_percentage >= 0.0 && response.progress_percentage <= 100.0);
        }
        Err(_) => {
            // Expected if collection doesn't exist
        }
    }
}

#[test]
fn test_graph_models_serialization() {
    // Test that graph models can be serialized/deserialized
    use vectorizer_sdk::models::graph::GraphNode;
    
    let node = GraphNode {
        id: "test_node".to_string(),
        node_type: "document".to_string(),
        metadata: std::collections::HashMap::new(),
    };
    
    let json = serde_json::to_string(&node).unwrap();
    let deserialized: GraphNode = serde_json::from_str(&json).unwrap();
    
    assert_eq!(node.id, deserialized.id);
    assert_eq!(node.node_type, deserialized.node_type);
}

File: sdks/csharp/Vectorizer.Tests/GraphTests.cs
using System;
using System.Threading.Tasks;
using Xunit;
using Vectorizer;
using Vectorizer.Models;

namespace Vectorizer.Tests
{
    public class GraphTests
    {
        private readonly VectorizerClient _client;

        public GraphTests()
        {
            _client = new VectorizerClient(new ClientConfig
            {
                BaseUrl = "http://localhost:15002"
            });
        }

        [Fact]
        public async Task ListGraphNodes_ShouldReturnNodes()
        {
            try
            {
                var result = await _client.ListGraphNodesAsync("test_collection");
                
                Assert.NotNull(result);
                Assert.True(result.Count >= 0);
                Assert.NotNull(result.Nodes);
            }
            catch
            {
                // Collection doesn't exist or graph not enabled - this is expected in test environment
            }
        }

        [Fact]
        public async Task GetGraphNeighbors_ShouldReturnNeighbors()
        {
            try
            {
                var result = await _client.GetGraphNeighborsAsync("test_collection", "test_node");
                
                Assert.NotNull(result);
                Assert.NotNull(result.Neighbors);
            }
            catch
            {
                // Collection/node doesn't exist - this is expected in test environment
            }
        }

        [Fact]
        public async Task FindRelatedNodes_ShouldReturnRelatedNodes()
        {
            try
            {
                var request = new FindRelatedRequest
                {
                    MaxHops = 2,
                    RelationshipType = "SIMILAR_TO"
                };
                
                var result = await _client.FindRelatedNodesAsync("test_collection", "test_node", request);
                
                Assert.NotNull(result);
                Assert.NotNull(result.Related);
            }
            catch
            {
                // Collection/node doesn't exist - this is expected in test environment
            }
        }

        [Fact]
        public async Task FindGraphPath_ShouldReturnPath()
        {
            try
            {
                var request = new FindPathRequest
                {
                    Collection = "test_collection",
                    Source = "node1",
                    Target = "node2"
                };
                
                var result = await _client.FindGraphPathAsync(request);
                
                Assert.NotNull(result);
                Assert.NotNull(result.Path);
                if (result.Found)
                {
                    Assert.NotEmpty(result.Path);
                }
            }
            catch
            {
                // Collection/nodes don't exist - this is expected in test environment
            }
        }

        [Fact]
        public async Task CreateGraphEdge_ShouldCreateEdge()
        {
            try
            {
                var request = new CreateEdgeRequest
                {
                    Collection = "test_collection",
                    Source = "node1",
                    Target = "node2",
                    RelationshipType = "SIMILAR_TO",
                    Weight = 0.85f
                };
                
                var result = await _client.CreateGraphEdgeAsync(request);
                
                Assert.NotNull(result);
                Assert.True(result.Success);
                Assert.NotEmpty(result.EdgeId);
            }
            catch
            {
                // Collection/nodes don't exist - this is expected in test environment
            }
        }

        [Fact]
        public async Task ListGraphEdges_ShouldReturnEdges()
        {
            try
            {
                var result = await _client.ListGraphEdgesAsync("test_collection");
                
                Assert.NotNull(result);
                Assert.True(result.Count >= 0);
                Assert.NotNull(result.Edges);
            }
            catch
            {
                // Collection doesn't exist - this is expected in test environment
            }
        }

        [Fact]
        public async Task DiscoverGraphEdges_ShouldDiscoverEdges()
        {
            try
            {
                var request = new DiscoverEdgesRequest
                {
                    SimilarityThreshold = 0.7f,
                    MaxPerNode = 10
                };
                
                var result = await _client.DiscoverGraphEdgesAsync("test_collection", request);
                
                Assert.NotNull(result);
                Assert.True(result.Success);
                Assert.True(result.EdgesCreated >= 0);
            }
            catch
            {
                // Collection doesn't exist - this is expected in test environment
            }
        }

        [Fact]
        public async Task GetGraphDiscoveryStatus_ShouldReturnStatus()
        {
            try
            {
                var result = await _client.GetGraphDiscoveryStatusAsync("test_collection");
                
                Assert.NotNull(result);
                Assert.True(result.TotalNodes >= 0);
                Assert.True(result.NodesWithEdges >= 0);
                Assert.True(result.TotalEdges >= 0);
                Assert.True(result.ProgressPercentage >= 0 && result.ProgressPercentage <= 100);
            }
            catch
            {
                // Collection doesn't exist - this is expected in test environment
            }
        }
    }
}

